native/snappy/src/lib.rs

```diff
@@ -0,0 +1,33 @@
+use std::io::{Read, Write};
+
+use rustler::{Binary, Env, NewBinary};
+use snap::read;
+
+fn bytes_to_binary<'env>(env: Env<'env>, bytes: &[u8]) -> Binary<'env> {
+    let mut binary = NewBinary::new(env, bytes.len());
+    // This cannot fail because bin size equals bytes len
+    binary.as_mut_slice().write_all(bytes).unwrap();
+    binary.into()
+}
+
+#[rustler::nif]
+fn compress<'env>(env: Env<'env>, bin: Binary) -> Result<Binary<'env>, String> {
+    let mut encoder = read::FrameEncoder::new(&bin[..]);
+    let mut buffer = Vec::with_capacity(bin.len());
+    encoder
+        .read_to_end(&mut buffer)
+        .map_err(|e| e.to_string())?;
+    Ok(bytes_to_binary(env, &buffer))
+}
+
+#[rustler::nif]
+fn decompress<'env>(env: Env<'env>, bin: Binary) -> Result<Binary<'env>, String> {
```

Comment:
very nice! BTW have you looked into the different types of Binary there is? https://docs.rs/rustler/latest/rustler/types/binary/index.html I don't completely understand the differences between them but just wanted to show this in case you haven't seen it before

Reply:
Yeah. AFAIK the `Binary` would be a read-only string that's owned by the calling process, the `OwnedBinary` is a read-write string that's owned by the NIF and isn't associated with a VM process (an "env"), and `NewBinary` is just a convenience, which works like an `OwnedBinary` (it's read-write) but already associated to a process.